src/com/cdd/bao/template/SchemaTree.java

```diff
@@ -138,7 +138,36 @@ public String toString()
 		}
 		return ancestors.toArray(new String[ancestors.size()]);
 	}
-	
+
+	// tack on value node to this schema tree, leaving the underlying assignment untouched
+	public void addNode(String parentURI, String label, String descr, String uri)
+	{
+		Node parent = tree.get(parentURI);
+
+		// assume flat list has already been calculated
+		Node node = new Node();
+		if (parent != null)
+		{
+			node.parent = parent;
+			node.depth = parent.depth + 1;
+			node.parentIndex = flat.indexOf(parent);
+
+			parent.children.add(node);
+			++parent.childCount;
+		}
+
+		node.uri = uri;
+		node.label = label;
+		node.descr = descr;
+		node.inSchema = false;
+		node.isExplicit = false;
+		tree.put(node.uri, node);
+
+		// resort after adding node to list
+		list.add(node);
+		list.sort((v1, v2) -> v1.label.compareToIgnoreCase(v2.label));
```

Comment:
You need to update `flat` by inserting at the correct position, and then sure the `parentIndex` for everything else is correct after the insertion.

Reply:
So, I thought about this, and my thinking was that new-node positioning might be a non-issue.
Can you confirm whether this is or is not the case?  Would we lose anything by merely appending
the new node to the end of FLAT array?

If we do need the usual positioning, then likely I would just clear FLAT, and call `SchemaTree.flattenBranch`
again.  In fact, I'm going to do that now, and wait for comments afterwards.

Reply:
It goes to some trouble to maintain that list and keep the indexes uptodate, so yes, it is important. The information is emitted to the client, which uses it for various tree-building exercises.

One approach that might make more sense is to keep `addNode` as one-at-a-time call, and just make sure that the tree objects are uptodate. Then add another method for rebuilding both the `flat` and `list` arrays (similar to the Vocab-based constructor). If it's only done once right after all of the new content has been added, there wouldn't be any performance problems.

That means the external code would have to do the parent checking, and implicitly add the content in an order that allows multiple nesting levels... but that should be easy enough.